lib/exename.cpp

```diff
@@ -16,58 +16,77 @@ limitations under the License.
-    const char *separator = cwd[cwdLen - 1] == '/' ? "" : "/";
+#if defined(__linux__) || defined(__FreeBSD__) || defined(__NetBSD__)
+    // Find the path of the executable.  We use a number of techniques that may fail or work on
+    // different systems, and take the first working one we find.  Fallback to not overriding the
+    // compiled-in installation path.
+    std::array<std::string, 4> paths = {
+        "/proc/self/exe",         // Linux
+        "/proc/curproc/file",     // FreeBSD
+        "/proc/curproc/exe",      // NetBSD
+        "/proc/self/path/a.out",  // others?
+    };
 
-    // Construct an absolute path. We're assuming that @relPath is relative to
-    // the current working directory.
-    int n = snprintf(output, N, "%s%s%s", cwd, separator, relPath);
-    BUG_CHECK(n >= 0, "Pathname too long");
+    for (const auto &path : paths) {
+        try {
+            // std::filesystem::canonical will throw on error if the path is invalid.
+            // It will also try to resolve symlinks.
+            std::filesystem::path exePath = std::filesystem::canonical(path);
+            strncpy(BUFFER.data(), exePath.c_str(), BUFFER.size() - 1);
+            BUFFER[BUFFER.size() - 1] = '\0';
+            return BUFFER.data();
+        } catch (const std::filesystem::filesystem_error &) {
+            // Ignore and try the next path.
+        }
+    }
+#elif defined(__APPLE__)
+    uint32_t size = static_cast<uint32_t>(BUFFER.size());
+    if (_NSGetExecutablePath(BUFFER.data(), &size) == 0) {
+        return (BUFFER.data());
+    }
+#elif defined(_WIN32)
+    // TODO: Do we need to support this?
+    DWORD size = GetModuleFileNameA(nullptr, BUFFER.data(), static_cast<DWORD>(BUFFER.size()));
+    if (size > 0 && size < BUFFER.size()) {
+        return (BUFFER.data());
+    }
+#endif
+    // If the above fails, try to convert argv0 to a path.
+    if (std::filesystem::exists(suggestedPath)) {
+        return suggestedPath;
+    }
+    return {};
 }
 
 const char *exename(const char *argv0) {
-    // Leave 1 extra char for the \0
-    static char buffer[PATH_MAX + 1];
-    if (buffer[0]) return buffer;  // done already
-    int len;
-    /* find the path of the executable.  We use a number of techniques that may fail
-     * or work on different systems, and take the first working one we find.  Fall
-     * back to not overriding the compiled-in installation path */
-    if ((len = readlink("/proc/self/exe", buffer, sizeof(buffer) - 1)) > 0 ||
-        (len = readlink("/proc/curproc/exe", buffer, sizeof(buffer) - 1)) > 0 ||
-        (len = readlink("/proc/curproc/file", buffer, sizeof(buffer) - 1)) > 0 ||
-        (len = readlink("/proc/self/path/a.out", buffer, sizeof(buffer) - 1)) > 0) {
-        buffer[len] = 0;
-    } else if (argv0 && argv0[0] == '/') {
-        snprintf(buffer, sizeof(buffer), "%s", argv0);
-    } else if (argv0 && strchr(argv0, '/')) {
-        convertToAbsPath(argv0, buffer);
-    } else if (getenv("_")) {
```

Comment:
Not sure about this use case. Does it work on MacOS? When is this needed?

Reply:
This should work if the compiler was invoked by bash, regardless of system.  Bash (and possibly other shells) set this envvar to the full, absolute path of the program being invoked when they call exec.  It is a useful fallback if argv0 is null or just a command name (that was looked up in $PATH).

Reply:
Use of `_` leads to issues that are very hard to reproduce / debug:
 - There are other shells besides `bash`
 - p4c could be invoked from some wrapper script and / or app. In such case `_` might or might not be present depending on how environment is sanitized. Even worse, it may contain path to wrapper, not to `p4c` itself

I spent quite some time trying to debug wrong include paths that only manifested under test suite run and were not reproducible when the exactly same command was run manually from the shell...